bin/cigar_parser.R

```diff
@@ -0,0 +1,1188 @@
+    res <- res %>% group_by(pos, nucleotide) %>% dplyr::summarise(count = sum(count)) ### to join counts of fw with rv alignment and avoid position and nucleotides duplicates
+    ### Get percentages
+    fperc <- res %>% group_by(pos) %>% transmute(percentage = round((count / sum(count))*100, 2))
+    fperc$nucleotide <- res$nucleotide
+    ### Return data frame with all positions and the percentage of each nucleotide by position
+    return(fperc)
+}
+
+#############
+### Template-based quantification
+#############
+templateCount <- function(ori_ref_file, new_ref_temp, template_bam, readsAlnInfo_collaps, sampleID, alnCompleteInfo_corrected, alnCompleteInfo_NOTcorrected){
+    ########
+    ### Function to get number of reads supporting template-based edits. The return will be a list with two values: count and kind of substitution
+    ########
+    #
+    #### Get cigar
+    param <- ScanBamParam(
+        flag=scanBamFlag(isUnmappedQuery = FALSE, isSecondaryAlignment = FALSE, isDuplicate = FALSE, isSupplementaryAlignment = FALSE), ## Get just primary alignments
+        what=c("cigar"))
+    bam=scanBam(template_bam, param=param)
+    temp_cigar=bam[[1]]$cigar[1]
+    if(is.na(temp_cigar)){
+        tempCount <- c(0, "no-changes")
+    } else {
+        #### Cigar length and type
+        cigar_lengths=explodeCigarOpLengths(temp_cigar) ##deconvolution of CIGAR field ie: M50I2M70 --> 50,2,70
+        cigar_types=explodeCigarOps(temp_cigar) ##deconvolution of CIGAR field ie: M50I2M70 --> M, I, M
+        #### Remove S from cigar (it happens when one of both arms of the template are longer than the reference)
+        if (cigar_types[[1]][1] == "S"){
+            s_slice <- cigar_lengths[[1]][1]
+            cigar_lengths[[1]] <- cigar_lengths[[1]][2:length(cigar_lengths[[1]])]
+            cigar_types[[1]] <- cigar_types[[1]][2:length(cigar_types[[1]])]
+            temp_cigar <- strsplit(temp_cigar, "S")[[1]][2]
+            if(cigar_types[[1]][length(cigar_types[[1]])] == "S"){
+                temp_cigar <- substr(temp_cigar, 1, nchar(temp_cigar) - nchar(cigar_lengths[[1]][length(cigar_lengths[[1]])]))
+                cigar_lengths[[1]] <- cigar_lengths[[1]][1:length(cigar_lengths[[1]])-1]
+                cigar_types[[1]] <- cigar_types[[1]][1:length(cigar_types[[1]])-1]
+            }
+        } else if(cigar_types[[1]][length(cigar_types[[1]])] == "S"){
+            s_slice <- 0
+            temp_cigar <- strsplit(temp_cigar, "S")[[1]][1]
+            temp_cigar <- substr(temp_cigar, 1, nchar(temp_cigar) - nchar(cigar_lengths[[1]][length(cigar_lengths[[1]])]))
+            cigar_lengths[[1]] <- cigar_lengths[[1]][1:length(cigar_lengths[[1]])-1]
+            cigar_types[[1]] <- cigar_types[[1]][1:length(cigar_types[[1]])-1]
+        } else {
+            s_slice <- 0
+        }
+        ##### Check if there is any read that looks like the template!
+        if(dim(readsAlnInfo_collaps %>% filter(cigar == temp_cigar))[1] == 0){
```

Comment:
## But before update by alignment position start
    cingLen_info <- explodeCigarOpLengths(readsAlnInfo_collaps$cigar)
    first_pos_len_info <- lapply( c(1:length(cingLen_info)), function(p){ cingLen_info[p][[1]][1] + readsAlnInfo_collaps$pos[p] - 1 } )
    all_pos_len_info <- lapply( c(1:length(cingLen_info)), function(l) { unlist( c(first_pos_len_info[l], explodeCigarOpLengths(readsAlnInfo_collaps$cigar[l])[[1]][2:length(explodeCigarOpLengths(readsAlnInfo_collaps$cigar[l])[[1]])])) } )
    temp_cigar_info=lapply( c(1: length(cingLen_info)), function(k) { paste(unlist( mapply(c, all_pos_len_info[k][[1]], explodeCigarOps(readsAlnInfo_collaps$cigar[k])[[1]], SIMPLIFY=FALSE)), collapse = "") }  )
    readsAlnInfo_collaps$cigar_pos <- unlist(temp_cigar_info)
    ### And now let's see
    if(dim(readsAlnInfo_collaps %>% filter(cigar_pos == temp_cigar))[1] == 0){